src/allocator_defrag.c

```diff
@@ -0,0 +1,379 @@
+        assert(je_res == 0);
+
+        // init bin specific fast query keys
+        je_res = binQueryHelperInitialization(&bin_info->info_keys, j);
+        assert(je_res == 0);
+
+        /* verify the reverse map of reg_size to bin index */
+        assert(jeSize2BinIndexLgQ3(bin_info->reg_size) == j);
+    }
+
+    // defrag is supported mark it to enable defrag queries
+    defrag_supported = 1;
+    return 0;
+}
+
+/* Total size of consumed meomry in unused regs in small bins (AKA external fragmentation).
+ * The function will refresh the epoch.
+ *
+ * return total fragmentation bytes
+ */
+unsigned long allocatorDefragGetFragSmallbins(void) {
+    assert(defrag_supported);
+    unsigned long frag = 0;
+    jeRefreshStats(&arena_bin_conf);
+    for (unsigned j = 0; j < arena_bin_conf.nbins; j++) {
+        size_t sz;
+        jeBinInfo *bin_info = &arena_bin_conf.bin_info[j];
+        jeBusage *bin_usage = &curr_usage[j];
+
+        /* Number of used regions in the bin */
+        sz = sizeof(size_t);
+        /* Number of current slabs in the bin */
+        jeQueryCtlInterface(&bin_info->info_keys.curr_regs, &bin_usage->curr_regs, &sz);
+        /* Number of current slabs in the bin */
+        jeQueryCtlInterface(&bin_info->info_keys.curr_slabs, &bin_usage->curr_slabs, &sz);
+        /* Number of non full slabs in the bin */
+        jeQueryCtlInterface(&bin_info->info_keys.nonfull_slabs, &bin_usage->curr_nonfull_slabs, &sz);
+
+        bin_usage->curr_full_slabs = bin_usage->curr_slabs - bin_usage->curr_nonfull_slabs;
+        /* Calculate the fragmentation bytes for the current bin and add it to the total. */
+        frag += ((bin_info->nregs * bin_usage->curr_slabs) - bin_usage->curr_regs) * bin_info->reg_size;
+    }
+    return frag;
+}
+
+/* Determines whether defragmentation should be performed on a pointer based on jemalloc information.
+ *
+ * bin_info Pointer to the bin information structure.
+ * bin_usage Pointer to the bin usage structure.
+ * nalloced Number of allocated regions in the bin.
+ *
+ * return 1 if defragmentation should be performed, 0 otherwise.
+ *
+ * This function checks the following conditions to determine if defragmentation should be performed:
+ * 1. If the number of allocated regions (nalloced) is equal to the total number of regions (bin_info->nregs),
+ *    defragmentation is not necessary as moving regions is guaranteed not to change the fragmentation ratio.
+ * 2. If the number of non-full slabs (bin_usage->curr_nonfull_slabs) is less than 2, defragmentation is not performed
+ *    because there is no other slab to move regions to.
+ * 3. If slab utilization < 'avg utilization'*1.125 [code 1.125 == (1000+UTILIZATION_THRESHOLD_FACTOR_MILI)/1000]
+ *    than we should defrag. This is aligned with previous je_defrag_hint implementation.
+ */
+static inline int makeDefragDecision(jeBinInfo *bin_info, jeBusage *bin_usage, unsigned long nalloced) {
```

Comment:
I'm unclear on the difference between `nalloced` and `bin_usage->curr_regs`.  Aren't these the same thing?

Reply:
No they are not. See the documentation. `bin_usage` represents bin level information, over all slabs in that bin. As such, `bin_usage->curr_regs` are the number of regions currently in used in the bin cross all slabs of that bin. `nalloced` is the number of regions allocated in the slab we want to make a decision about.